[GTK][WPE] Pass platform argument to test262 runner

philn opened the pull request:
#### 0d823991a9c000ecebedabc3717440953eb41f86
<pre>
[GTK][WPE] Pass platform argument to test262 runner
<a href="https://bugs.webkit.org/show_bug.cgi?id=302549">https://bugs.webkit.org/show_bug.cgi?id=302549</a>
&lt;<a href="https://rdar.apple.com/problem/165232128">rdar://problem/165232128</a>&gt;

Reviewed by Carlos Alberto Lopez Perez.

By passing the platform argument to the runner, the product dir can be reliably determined and
cached before processing CLI options. Without this change the runner won&apos;t find jsc on WPE/GTK
builds, in fact this has been broken since the migration to the new Linux SDK.

Canonical link: <a href="https://commits.webkit.org/303610@main">https://commits.webkit.org/303610@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34fbd0b5651eef898e7965b1894a69de91cca9b5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132992 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5493 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44088 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140528 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134862 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5925 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5356 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101698 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135938 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119163 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82497 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1659 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83762 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113097 "Passed tests") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143181 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5162 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37862 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110076 "Found 1 new test failure: imported/blink/fast/pagination/viewport-y-horizontal-bt-rtl.html (failure)") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/132420 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5244 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4420 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110257 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3958 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115425 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58733 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20604 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5216 "Built successfully") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5057 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5306 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/5174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (changes requested):
It seems that test262-runner is requiring platform parameter, this is not great since it worked by gussing the JSC bin path before, since we use this command locally too, not just for bot. Can you align the behavior to run-jsc-stress-tests? (which guesses the path already)

mcatanzaro commented on Tools/Scripts/test262/Runner.pm:
```diff
@@ -160,6 +160,10 @@ sub processCLI {
     my $specifiedResultsFile;
     my $specifiedExpectationsFile;
 
+    # Cache the product dir. Do this before processing CLI options because it implicitely processes
```
```suggestion
    # Cache the product dir. Do this before processing CLI options because it implicitly processes
```

Constellation commented on Tools/CISupport/build-webkit-org/steps_unittest.py:
```diff
@@ -2439,7 +2439,7 @@ def test_success(self):
                         timeout=7200,
                         log_environ=True,
-                        command=['/bin/bash', '--posix', '-o', 'pipefail', '-c', 'perl Tools/Scripts/test262-runner --verbose --release 2>&1 | python3 Tools/Scripts/filter-test-logs test262'],
+                        command=['/bin/bash', '--posix', '-o', 'pipefail', '-c', 'perl Tools/Scripts/test262-runner --verbose --release --mac-sonoma 2>&1 | python3 Tools/Scripts/filter-test-logs test262'],
```
Let's remove this if it is not necessary :) This encoded arch name would make update complicated.

philn replied:
With this patch on macOS if you call `test262-runner` without platform argument, does it work?

If so, then this patch is harmless and I would like to avoid WPE/GTK-specific buildbot steps for this.

Constellation commented on Tools/CISupport/build-webkit-org/steps_unittest.py:
```diff
@@ -2454,7 +2454,7 @@ def test_failure(self):
                         timeout=7200,
                         log_environ=True,
-                        command=['/bin/bash', '--posix', '-o', 'pipefail', '-c', 'perl Tools/Scripts/test262-runner --verbose --debug 2>&1 | python3 Tools/Scripts/filter-test-logs test262'],
+                        command=['/bin/bash', '--posix', '-o', 'pipefail', '-c', 'perl Tools/Scripts/test262-runner --verbose --debug --mac-sonoma 2>&1 | python3 Tools/Scripts/filter-test-logs test262'],
```
Ditto

Constellation reviewed (commented):
> The argument is required only for GTK and WPE, because they use a per-port product dir, which apparently isn't the case for the other ports.

How about JSCOnly port?

clopez commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -824,7 +824,7 @@ class RunTest262Tests(TestWithFailureCount, CustomFlagsMixin, ShellMixin):
     description = ["test262-tests running"]
     descriptionDone = ["test262-tests"]
     failedTestsFormatString = "%d Test262 test%s failed"
-    command = ["perl", "Tools/Scripts/test262-runner", "--verbose", WithProperties("--%(configuration)s")]
+    command = ["perl", "Tools/Scripts/test262-runner", "--verbose", WithProperties("--%(configuration)s"), WithProperties("--%(fullPlatform)s")]
     test_summary_re = re.compile(r'^\! NEW FAIL')
 
     def __init__(self, *args, **kwargs):
```
This is not needed.

You have to fix below the code so it executes the `self.appendCustomBuildFlags(self.getProperty('platform'), self.getProperty('fullPlatform'))` before the `filter_command = ' '.join(self.command) + ' 2>&1 | python3 Tools/Scripts/filter-test-logs test262'`

Look at what the bot is executing: https://build.webkit.org/#/builders/64/builds/247581/steps/15/logs/stdio

```
argv: [b'/bin/bash', b'--posix', b'-o', b'pipefail', b'-c', b'perl Tools/Scripts/test262-runner --verbose --debug 2>&1 | python3 Tools/Scripts/filter-test-logs test262', b'--gtk']
```

is passing the `--gtk` flag after `| python3 Tools/Scripts/filter-test-logs test262'`

The right patch would be:

```diff
--- a/Tools/CISupport/build-webkit-org/steps.py
+++ b/Tools/CISupport/build-webkit-org/steps.py
@@ -831,13 +831,13 @@ class RunTest262Tests(TestWithFailureCount, CustomFlagsMixin, ShellMixin):
         super().__init__(*args, timeout=2 * 60 * 60, **kwargs)
 
     def run(self):
+        self.appendCustomBuildFlags(self.getProperty('platform'), self.getProperty('fullPlatform'))
         filter_command = ' '.join(self.command) + ' 2>&1 | python3 Tools/Scripts/filter-test-logs test262'
         self.command = self.shell_command(filter_command)
 
         self.log_observer = ParseByLineLogObserver(self.parseOutputLine)
         self.addLogObserver('stdio', self.log_observer)
         self.failedTestCount = 0
-        self.appendCustomBuildFlags(self.getProperty('platform'), self.getProperty('fullPlatform'))
 
         steps_to_add = [
             GenerateS3URL(
```

Also notice that `self.appendCustomBuildFlags` doesn't appends this flag for mac.

This looks like a regression introduced in https://commits.webkit.org/299905@main

clopez reviewed (changes requested):
the buildbot steps code needs to be reworked, check my comment above

philn committed:
[GTK][WPE] Pass platform argument to test262 runner
https://bugs.webkit.org/show_bug.cgi?id=302549
<rdar://problem/165232128>

Reviewed by Carlos Alberto Lopez Perez.

By passing the platform argument to the runner, the product dir can be reliably determined and
cached before processing CLI options. Without this change the runner won't find jsc on WPE/GTK
builds, in fact this has been broken since the migration to the new Linux SDK.

Canonical link: https://commits.webkit.org/303610@main
